README.md
# messagePack_dart

[![pub package](https://img.shields.io/pub/v/message_pack_dart.svg)](https://pub.dartlang.org/packages/message_pack_dart)

A MessagePack implementation for dart/Flutter [MessagePack](https://msgpack.org).

MessagePack is an efficient binary serialization format. It lets you exchange data among multiple languages like JSON. But it's faster and smaller. Small integers are encoded into a single byte, and typical short strings require only one extra byte in addition to the strings themselves.



CHANGELOG.md
## [2.0.1] - 2022-07-11

* Fix readme

## [2.0.0] - 2022-07-11

* Add WEB support
* Format dart
* Improve code structure

## [1.0.0] - 2021-03-11

* Migrated to null safety (thanks RootSoft)

## [0.0.7] - 2020-03-14

* Fix wrong length when writing ext8 (0xc7)

## [0.0.6] - 2019-09-20

* Accept any iterable when serializing, not just List
* Accept ByteData when serializing (will be deserialized as Uint8List)

## [0.0.5] - 2019-05-19

* Changed return value from `List<int>` to `Uint8List`.


test/message_pack_dart_test.dart
import 'dart:typed_data';

import 'package:message_pack_dart/message_pack_dart.dart';
import 'package:test/test.dart';

//
// Tests taken from msgpack2 (https://github.com/butlermatt/msgpack2)
//

var isString = predicate((e) => e is String, 'is a String');
var isInt = predicate((e) => e is int, 'is an int');
var isMap = predicate((e) => e is Map, 'is a Map');
var isList = predicate((e) => e is List, 'is a List');

void main() {
  test("Test Pack null", packNull);

  group("Test Pack Boolean", () {
    test("Pack boolean false", packFalse);
    test("Pack boolean true", packTrue);
  });

  group("Test Pack Ints", () {
    test("Pack Positive FixInt", packPositiveFixInt);
    test("Pack Negative FixInt", packFixedNegative);
    test("Pack Uint8", packUint8);
    test("Pack Uint16", packUint16);
    test("Pack Uint32", packUint32);
    test("Pack Uint64", packUint64);
    test("Pack Int8", packInt8);
    test("Pack Int16", packInt16);
    test("Pack Int32", packInt32);
    test("Pack Int64", packInt64);
  });

  group("Test Pack Floats", () {
    test("Pack Float32", packFloat32);
    test("Pack Float64 (double)", packDouble);
  });

  test("Pack 5-character string", packString5);
  test("Pack 22-character string", packString22);
  test("Pack 256-character string", packString256);
  test("Pack string array", packStringArray);
  test("Pack int-to-string map", packIntToStringMap);

  group("Test Pack Binary", () {
    test("Pack Bin8", packBin8);
    test("Pack Bin16", packBin16);
    test("Pack Bin32", packBin32);
    test("Pack ByteData", packByteData);
  });

  test("Test Unpack Null", unpackNull);

  group("Test Unpack boolean", () {
    test("Unpack boolean false", unpackFalse);
    test("Unpack boolean true", unpackTrue);
  });

  group("Test Unpack Ints", () {
    test("Unpack Positive FixInt", unpackPositiveFixInt);
    test("Unpack Negative FixInt", unpackNegativeFixInt);
    test("Unpack Uint8", unpackUint8);
    test("Unpack Uint16", unpackUint16);
    test("Unpack Uint32", unpackUint32);
    test("Unpack Uint64", unpackUint64);
    test("Unpack Int8", unpackInt8);
    test("Unpack Int16", unpackInt16);
    test("Unpack Int32", unpackInt32);
    test("Unpack Int64", unpackInt64);
  });

  group("Test Unpack Floats", () {
    test("Unpack Float32", unpackFloat32);
    test("Unpack Float64 (double)", unpackDouble);
  });

  test("Unpack 5-character string", unpackString5);
  test("Unpack 22-character string", unpackString22);
  test("Unpack 256-character string", unpackString256);
  test("Unpack string array", unpackStringArray);
  test("Unpack int-to-string map", unpackIntToStringMap);

  group("Test Large Array and Map", () {
    test("Large Array", largeArray);
    test("Very Large Array", veryLargeArray);
    test("Large Map", largeMap);
    test("Very Large Map", veryLargeMap);
  });
}

void largeArray() {
  final list = <String>[];
  for (int i = 0; i < 16; ++i) {
    list.add("Item $i");
  }

  final serialized = serialize(list);
  List deserialized = deserialize(serialized);
  expect(deserialized, list);
}

void veryLargeArray() {
  final list = <String>[];
  for (int i = 0; i < 65536; ++i) {
    list.add("Item $i");
  }

  final serialized = serialize(list);
  List deserialized = deserialize(serialized);
  expect(deserialized, list);
}

void largeMap() {
  final map = Map<int, String>();
  for (int i = 0; i < 16; ++i) {
    map[i] = "Item $i";
  }
  final serialized = serialize(map);
  Map deserialized = deserialize(serialized);
  expect(deserialized, map);
}

void veryLargeMap() {
  final map = Map<int, String>();
  for (int i = 0; i < 65536; ++i) {
    map[i] = "Item $i";
  }
  final serialized = serialize(map);
  Map deserialized = deserialize(serialized);
  expect(deserialized, map);
}

void packNull() {
  List<int> encoded = serialize(null);
  expect(encoded, orderedEquals([0xc0]));
}

void packFalse() {
  List<int> encoded = serialize(false);
  expect(encoded, orderedEquals([0xc2]));
}

void packTrue() {
  List<int> encoded = serialize(true);
  expect(encoded, orderedEquals([0xc3]));
}

void packPositiveFixInt() {
  List<int> encoded = serialize(1);
  expect(encoded, orderedEquals([1]));
}

void packFixedNegative() {
  List<int> encoded = serialize(-16);
  expect(encoded, orderedEquals([240]));
}

void packUint8() {
  List<int> encoded = serialize(128);
  expect(encoded, orderedEquals([204, 128]));
}

void packUint16() {
  List<int> encoded = serialize(32768);
  expect(encoded, orderedEquals([205, 128, 0]));
}

void packUint32() {
  List<int> encoded = serialize(2147483648);
  expect(encoded, orderedEquals([206, 128, 0, 0, 0]));
}

void packUint64() {
  List<int> encoded = serialize(9223372036854775807);
  expect(encoded, orderedEquals([207, 127, 255, 255, 255, 255, 255, 255, 255]));
}

void packInt8() {
  List<int> encoded = serialize(-128);
  expect(encoded, orderedEquals([208, 128]));
}

void packInt16() {
  List<int> encoded = serialize(-32768);
  expect(encoded, orderedEquals([209, 128, 0]));
}

void packInt32() {
  List<int> encoded = serialize(-2147483648);
  expect(encoded, orderedEquals([210, 128, 0, 0, 0]));
}

void packInt64() {
  List<int> encoded = serialize(-9223372036854775808);
  expect(encoded, orderedEquals([211, 128, 0, 0, 0, 0, 0, 0, 0]));
}

void packFloat32() {
  List<int> encoded = serialize(Float(3.14));
  expect(encoded, orderedEquals([202, 64, 72, 245, 195]));
}

void packDouble() {
  List<int> encoded = serialize(3.14);
  expect(encoded,
      orderedEquals([0xcb, 0x40, 0x09, 0x1e, 0xb8, 0x51, 0xeb, 0x85, 0x1f]));
}

void packString5() {
  List<int> encoded = serialize("hello");
  expect(encoded, orderedEquals([165, 104, 101, 108, 108, 111]));
}

void packString22() {
  List<int> encoded = serialize("hello there, everyone!");
  expect(
      encoded,
      orderedEquals([
        182,
        104,
        101,
        108,
        108,
        111,
        32,
        116,
        104,
        101,
        114,
        101,
        44,
        32,
        101,
        118,
        101,
        114,
        121,
        111,
        110,
        101,
        33
      ]));
}

void packString256() {
  List<int> encoded = serialize(
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA");
  expect(encoded, hasLength(259));
  expect(encoded.sublist(0, 3), orderedEquals([218, 1, 0]));
  expect(encoded.sublist(3, 259), everyElement(65));
}

void packBin8() {
  var data = Uint8List.fromList(List.filled(32, 65));
  List<int> encoded = serialize(data);
  expect(encoded.length, equals(34));
  expect(encoded.getRange(0, 2), orderedEquals([0xc4, 32]));
  expect(encoded.getRange(2, encoded.length), orderedEquals(data));
}

void packBin16() {
  var data = Uint8List.fromList(List.filled(256, 65));
  List<int> encoded = serialize(data);
  expect(encoded.length, equals(256 + 3));
  expect(encoded.getRange(0, 3), orderedEquals([0xc5, 1, 0]));
  expect(encoded.getRange(3, encoded.length), orderedEquals(data));
}

void packBin32() {
  var data = Uint8List.fromList(List.filled(65536, 65));
  List<int> encoded = serialize(data);
  expect(encoded.length, equals(65536 + 5));
  expect(encoded.getRange(0, 5), orderedEquals([0xc6, 0, 1, 0, 0]));
  expect(encoded.getRange(5, encoded.length), orderedEquals(data));
}

void packByteData() {
  var data = ByteData.view(Uint8List.fromList(List.filled(32, 65)).buffer);
  List<int> encoded = serialize(data);
  expect(encoded.length, equals(34));
  expect(encoded.getRange(0, 2), orderedEquals([0xc4, 32]));
  expect(encoded.getRange(2, encoded.length),
      orderedEquals(data.buffer.asUint8List()));
}

void packStringArray() {
  List<int> encoded = serialize(["one", "two", "three"]);
  expect(
      encoded,
      orderedEquals([
        147,
        163,
        111,
        110,
        101,
        163,
        116,
        119,
        111,
        165,
        116,
        104,
        114,
        101,
        101
      ]));
}

void packIntToStringMap() {
  List<int> encoded = serialize({1: "one", 2: "two"});
  expect(encoded,
      orderedEquals([130, 1, 163, 111, 110, 101, 2, 163, 116, 119, 111]));
}

// Test unpacking
void unpackNull() {
  Uint8List data = new Uint8List.fromList([0xc0]);
  var value = deserialize(data);
  expect(value, isNull);
}

void unpackFalse() {
  Uint8List data = Uint8List.fromList([0xc2]);
  var value = deserialize(data);
  expect(value, isFalse);
}

void unpackTrue() {
  Uint8List data = Uint8List.fromList([0xc3]);
  var value = deserialize(data);
  expect(value, isTrue);
}

void unpackString5() {
  Uint8List data = new Uint8List.fromList([165, 104, 101, 108, 108, 111]);
  var value = deserialize(data);
  expect(value, isString);
  expect(value, equals("hello"));
}

void unpackString22() {
  Uint8List data = new Uint8List.fromList([
    182,
    104,
    101,
    108,
    108,
    111,
    32,
    116,
    104,
    101,
    114,
    101,
    44,
    32,
    101,
    118,
    101,
    114,
    121,
    111,
    110,
    101,
    33
  ]);
  var value = deserialize(data);
  expect(value, isString);
  expect(value, equals("hello there, everyone!"));
}

void unpackPositiveFixInt() {
  Uint8List data = Uint8List.fromList([1]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(1));
}

void unpackNegativeFixInt() {
  Uint8List data = Uint8List.fromList([240]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(-16));
}

void unpackUint8() {
  Uint8List data = Uint8List.fromList([204, 128]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(128));
}

void unpackUint16() {
  Uint8List data = Uint8List.fromList([205, 128, 0]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(32768));
}

void unpackUint32() {
  Uint8List data = Uint8List.fromList([206, 128, 0, 0, 0]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(2147483648));
}

void unpackUint64() {
  // Dart 2 doesn't support true Uint64 without using BigInt
  Uint8List data =
      Uint8List.fromList([207, 127, 255, 255, 255, 255, 255, 255, 255]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(9223372036854775807));
}

void unpackInt8() {
  Uint8List data = Uint8List.fromList([208, 128]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(-128));
}

void unpackInt16() {
  Uint8List data = Uint8List.fromList([209, 128, 0]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(-32768));
  data = Uint8List.fromList([0xd1, 0x04, 0xd2]);
  print(deserialize(data));
}

void unpackInt32() {
  Uint8List data = Uint8List.fromList([210, 128, 0, 0, 0]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(-2147483648));
}

void unpackInt64() {
  Uint8List data = Uint8List.fromList([211, 128, 0, 0, 0, 0, 0, 0, 0]);
  var value = deserialize(data);
  expect(value, isInt);
  expect(value, equals(-9223372036854775808));
}

void unpackFloat32() {
  Uint8List data = Uint8List.fromList([202, 64, 72, 245, 195]);
  var value = deserialize(data);
  expect((value as double).toStringAsPrecision(3), equals('3.14'));
}

void unpackDouble() {
  Uint8List data = Uint8List.fromList(
      [0xcb, 0x40, 0x09, 0x1e, 0xb8, 0x51, 0xeb, 0x85, 0x1f]);
  var value = deserialize(data);
  expect(value, equals(3.14));
}

void unpackString256() {
  Uint8List data =
      new Uint8List.fromList([218, 1, 0]..addAll(new List.filled(256, 65)));
  var value = deserialize(data);
  expect(value, isString);
  expect(
      value,
      equals(
          "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"));
}

void unpackStringArray() {
  Uint8List data = new Uint8List.fromList([
    147,
    163,
    111,
    110,
    101,
    163,
    116,
    119,
    111,
    165,
    116,
    104,
    114,
    101,
    101
  ]);
  var value = deserialize(data);
  expect(value, isList);
  expect(value, orderedEquals(["one", "two", "three"]));
}

void unpackIntToStringMap() {
  Uint8List data = new Uint8List.fromList(
      [130, 1, 163, 111, 110, 101, 2, 163, 116, 119, 111]);
  var value = deserialize(data);
  expect(value, isMap);
  expect(value[1], equals("one"));
  expect(value[2], equals("two"));
}

void unpackSmallDateTime() {
  var data = <int>[0xd7, 0xff, 0, 0, 0, 0, 0, 0, 0, 0];
  var value = deserialize(Uint8List.fromList(data));
  expect(value, equals(DateTime.fromMillisecondsSinceEpoch(0)));
  data = <int>[0xd7, 0xff, 47, 175, 8, 0, 91, 124, 180, 16];
  value = deserialize(Uint8List.fromList(data));
  expect((value as DateTime).toUtc(),
      equals(DateTime.utc(2018, 8, 22, 0, 56, 56, 200)));
}

void unpackPastDate() {
  var data = <int>[
    0xc7,
    12,
    0xff,
    29,
    205,
    101,
    0,
    255,
    255,
    255,
    255,
    184,
    204,
    121,
    158
  ];

  var value = deserialize(Uint8List.fromList(data)) as DateTime;
  expect(value.toUtc(), equals(DateTime.utc(1932, 2, 24, 1, 53, 45, 500)));

  data = <int>[
    199,
    12,
    255,
    0,
    0,
    0,
    0,
    255,
    255,
    255,
    255,
    255,
    255,
    248,
    248
  ];
  value = deserialize(Uint8List.fromList(data));
  expect(value.toUtc(), equals(DateTime.utc(1969, 12, 31, 23, 30)));
}


lib/src/data_writer.dart
part of messagePack_dart;

final int _kScratchSizeInitial = 64;
final int _kScratchSizeRegular = 1024;

class DataWriter {
  Uint8List? _scratchBuffer;
  ByteData? _scratchData;
  int _scratchOffset = 0;

  void writeUint8(int i) {
    _ensureSize(1);
    _scratchData?.setUint8(_scratchOffset, i);
    _scratchOffset += 1;
  }

  void writeInt8(int i) {
    _ensureSize(1);
    _scratchData?.setInt8(_scratchOffset, i);
    _scratchOffset += 1;
  }

  void writeUint16(int i, [Endian endian = Endian.big]) {
    _ensureSize(2);
    _scratchData?.setUint16(_scratchOffset, i, endian);
    _scratchOffset += 2;
  }

  void writeInt16(int i, [Endian endian = Endian.big]) {
    _ensureSize(2);
    _scratchData?.setInt16(_scratchOffset, i, endian);
    _scratchOffset += 2;
  }

  void writeUint32(int i, [Endian endian = Endian.big]) {
    _ensureSize(4);
    _scratchData?.setUint32(_scratchOffset, i, endian);
    _scratchOffset += 4;
  }

  void writeInt32(int i, [Endian endian = Endian.big]) {
    _ensureSize(4);
    _scratchData?.setInt32(_scratchOffset, i, endian);
    _scratchOffset += 4;
  }

  void writeUint64(int i, [Endian endian = Endian.big]) {
    _ensureSize(8);
    _scratchData?.setUint64(_scratchOffset, i, endian);
    _scratchOffset += 8;
  }

  void writeInt64(int i, [Endian endian = Endian.big]) {
    _ensureSize(8);
    _scratchData?.setInt64(_scratchOffset, i, endian);
    _scratchOffset += 8;
  }

  void writeFloat32(double f, [Endian endian = Endian.big]) {
    _ensureSize(4);
    _scratchData?.setFloat32(_scratchOffset, f, endian);
    _scratchOffset += 4;
  }

  void writeFloat64(double f, [Endian endian = Endian.big]) {
    _ensureSize(8);
    _scratchData?.setFloat64(_scratchOffset, f, endian);
    _scratchOffset += 8;
  }

  // The list may be retained until takeBytes is called
  void writeBytes(List<int> bytes) {
    final length = bytes.length;
    if (length == 0) {
      return;
    }
    _ensureSize(length);
    if (_scratchOffset == 0) {
      // we can add it directly
      _builder.add(bytes);
    } else {
      // there is enough room in _scratchBuffer, otherwise _ensureSize
      // would have added _scratchBuffer to _builder and _scratchOffset would
      // be 0
      if (bytes is Uint8List) {
        _scratchBuffer?.setRange(
            _scratchOffset, _scratchOffset + length, bytes);
      } else {
        for (int i = 0; i < length; i++) {
          _scratchBuffer?[_scratchOffset + i] = bytes[i];
        }
      }
      _scratchOffset += length;
    }
  }

  Uint8List takeBytes() {
    if (_builder.isEmpty) {
      // Just take scratch data
      final res = Uint8List.view(
        _scratchBuffer!.buffer,
        _scratchBuffer!.offsetInBytes,
        _scratchOffset,
      );
      _scratchOffset = 0;
      _scratchBuffer = null;
      _scratchData = null;
      return res;
    } else {
      _appendScratchBuffer();
      return _builder.takeBytes();
    }
  }

  void _ensureSize(int size) {
    if (_scratchBuffer == null) {
      // start with small scratch buffer, expand to regular later if needed
      _scratchBuffer = Uint8List(_kScratchSizeInitial);
      _scratchData =
          ByteData.view(_scratchBuffer!.buffer, _scratchBuffer!.offsetInBytes);
    }
    final remaining = _scratchBuffer!.length - _scratchOffset;
    if (remaining < size) {
      _appendScratchBuffer();
    }
  }

  void _appendScratchBuffer() {
    if (_scratchOffset > 0) {
      if (_builder.isEmpty) {
        // We're still on small scratch buffer, move it to _builder
        // and create regular one
        _builder.add(Uint8List.view(
          _scratchBuffer!.buffer,
          _scratchBuffer!.offsetInBytes,
          _scratchOffset,
        ));
        _scratchBuffer = Uint8List(_kScratchSizeRegular);
        _scratchData = ByteData.view(
            _scratchBuffer!.buffer, _scratchBuffer!.offsetInBytes);
      } else {
        _builder.add(
          Uint8List.fromList(
            Uint8List.view(
              _scratchBuffer!.buffer,
              _scratchBuffer!.offsetInBytes,
              _scratchOffset,
            ),
          ),
        );
      }
      _scratchOffset = 0;
    }
  }

  final _builder = BytesBuilder(copy: false);
}


lib/src/deserializer.dart
part of messagePack_dart;

abstract class ExtDecoder {
  dynamic decodeObject(int extType, Uint8List data);
}

class Deserializer {
  final ExtDecoder? _extDecoder;
  final codec = Utf8Codec();
  final Uint8List _list;
  final ByteData _data;
  int _offset = 0;

  Deserializer(
    Uint8List list, {
    ExtDecoder? extDecoder,
    this.copyBinaryData = false,
  })  : _list = list,
        _data = ByteData.view(list.buffer, list.offsetInBytes),
        _extDecoder = extDecoder;

  /// If false, decoded binary data buffers will reference underlying input
  /// buffer and thus may change when the content of input buffer changes.
  ///
  /// If true, decoded buffers are copies and the underlying input buffer is
  /// free to change after decoding.
  final bool copyBinaryData;

  dynamic decode() {
    final u = _list[_offset++];
    if (u <= 127) {
      return u;
    } else if ((u & 0xE0) == 0xE0) {
      // negative small integer
      return u - 256;
    } else if ((u & 0xE0) == 0xA0) {
      return _readString(u & 0x1F);
    } else if ((u & 0xF0) == 0x90) {
      return _readArray(u & 0xF);
    } else if ((u & 0xF0) == 0x80) {
      return _readMap(u & 0xF);
    }
    switch (u) {
      case 0xc0:
        return null;
      case 0xc2:
        return false;
      case 0xc3:
        return true;
      case 0xcc:
        return _readUInt8();
      case 0xcd:
        return _readUInt16();
      case 0xce:
        return _readUInt32();
      case 0xcf:
        return _readUInt64();
      case 0xd0:
        return _readInt8();
      case 0xd1:
        return _readInt16();
      case 0xd2:
        return _readInt32();
      case 0xd3:
        return _readInt64();
      case 0xca:
        return _readFloat();
      case 0xcb:
        return _readDouble();
      case 0xd9:
        return _readString(_readUInt8());
      case 0xda:
        return _readString(_readUInt16());
      case 0xdb:
        return _readString(_readUInt32());
      case 0xc4:
        return _readBuffer(_readUInt8());
      case 0xc5:
        return _readBuffer(_readUInt16());
      case 0xc6:
        return _readBuffer(_readUInt32());
      case 0xdc:
        return _readArray(_readUInt16());
      case 0xdd:
        return _readArray(_readUInt32());
      case 0xde:
        return _readMap(_readUInt16());
      case 0xdf:
        return _readMap(_readUInt32());
      case 0xd4:
        return _readExt(1);
      case 0xd5:
        return _readExt(2);
      case 0xd6:
        return _readExt(4);
      case 0xd7:
        return _readExt(8);
      case 0xd8:
        return _readExt(16);
      case 0xc7:
        return _readExt(_readUInt8());
      case 0xc8:
        return _readExt(_readUInt16());
      case 0xc9:
        return _readExt(_readUInt32());
      default:
        throw FormatError("Invalid MessagePack format");
    }
  }

  int _readInt8() {
    return _data.getInt8(_offset++);
  }

  int _readUInt8() {
    return _data.getUint8(_offset++);
  }

  int _readUInt16() {
    final res = _data.getUint16(_offset);
    _offset += 2;
    return res;
  }

  int _readInt16() {
    final res = _data.getInt16(_offset);
    _offset += 2;
    return res;
  }

  int _readUInt32() {
    final res = _data.getUint32(_offset);
    _offset += 4;
    return res;
  }

  int _readInt32() {
    final res = _data.getInt32(_offset);
    _offset += 4;
    return res;
  }

  int _readUInt64() {
    final res = _data.getUint64(_offset);
    _offset += 8;
    return res;
  }

  int _readInt64() {
    final res = _data.getInt64(_offset);
    _offset += 8;
    return res;
  }

  double _readFloat() {
    final res = _data.getFloat32(_offset);
    _offset += 4;
    return res;
  }

  double _readDouble() {
    final res = _data.getFloat64(_offset);
    _offset += 8;
    return res;
  }

  Uint8List _readBuffer(int length) {
    final res =
        Uint8List.view(_list.buffer, _list.offsetInBytes + _offset, length);
    _offset += length;
    return copyBinaryData ? Uint8List.fromList(res) : res;
  }

  String _readString(int length) {
    final list = _readBuffer(length);
    final len = list.length;
    for (int i = 0; i < len; ++i) {
      if (list[i] > 127) {
        return codec.decode(list);
      }
    }
    return String.fromCharCodes(list);
  }

  List _readArray(int length) {
    final res = List<dynamic>.filled(length, null, growable: false);
    for (int i = 0; i < length; ++i) {
      res[i] = decode();
    }
    return res;
  }

  Map _readMap(int length) {
    final res = Map();
    while (length > 0) {
      res[decode()] = decode();
      --length;
    }
    return res;
  }

  dynamic _readExt(int length) {
    final extType = _readUInt8();
    final data = _readBuffer(length);
    return _extDecoder?.decodeObject(extType, data);
  }
}


lib/src/serializer.dart
part of messagePack_dart;

abstract class ExtEncoder {
  // Return null if object can't be encoded
  int extTypeForObject(dynamic object);

  Uint8List encodeObject(dynamic object);
}

class Float {
  Float(this.value);

  final double value;

  @override
  String toString() => value.toString();
}

class Serializer {
  final _codec = Utf8Codec();
  final DataWriter _writer;
  final ExtEncoder? _extEncoder;

  Serializer({
    DataWriter? dataWriter,
    ExtEncoder? extEncoder,
  })  : _writer = dataWriter ?? DataWriter(),
        _extEncoder = extEncoder;

  void encode(dynamic d) {
    if (d == null) return _writer.writeUint8(0xc0);
    if (d is bool) return _writer.writeUint8(d == true ? 0xc3 : 0xc2);
    if (d is int) return d >= 0 ? _writePositiveInt(d) : _writeNegativeInt(d);
    if (d is Float) return _writeFloat(d);
    if (d is double) return _writeDouble(d);
    if (d is String) return _writeString(d);
    if (d is Uint8List) return _writeBinary(d);
    if (d is Iterable) return _writeIterable(d);
    if (d is ByteData)
      return _writeBinary(
          d.buffer.asUint8List(d.offsetInBytes, d.lengthInBytes));
    if (d is Map) return _writeMap(d);
    if (_extEncoder != null && _writeExt(d)) {
      return;
    }
    throw FormatError("Don't know how to serialize $d");
  }

  Uint8List takeBytes() {
    return _writer.takeBytes();
  }

  void _writeNegativeInt(int n) {
    if (n >= -32) {
      this._writer.writeInt8(n);
    } else if (n >= -128) {
      this._writer.writeUint8(0xd0);
      this._writer.writeInt8(n);
    } else if (n >= -32768) {
      this._writer.writeUint8(0xd1);
      this._writer.writeInt16(n);
    } else if (n >= -2147483648) {
      this._writer.writeUint8(0xd2);
      this._writer.writeInt32(n);
    } else {
      this._writer.writeUint8(0xd3);
      this._writer.writeInt64(n);
    }
  }

  void _writePositiveInt(int n) {
    if (n <= 127) {
      this._writer.writeUint8(n);
    } else if (n <= 0xFF) {
      this._writer.writeUint8(0xcc);
      this._writer.writeUint8(n);
    } else if (n <= 0xFFFF) {
      this._writer.writeUint8(0xcd);
      this._writer.writeUint16(n);
    } else if (n <= 0xFFFFFFFF) {
      this._writer.writeUint8(0xce);
      this._writer.writeUint32(n);
    } else {
      this._writer.writeUint8(0xcf);
      this._writer.writeUint64(n);
    }
  }

  void _writeFloat(Float n) {
    _writer.writeUint8(0xca);
    _writer.writeFloat32(n.value);
  }

  void _writeDouble(double n) {
    _writer.writeUint8(0xcb);
    _writer.writeFloat64(n);
  }

  void _writeString(String s) {
    final encoded = _codec.encode(s);
    final length = encoded.length;
    if (length <= 31) {
      _writer.writeUint8(0xA0 | length);
    } else if (length <= 0xFF) {
      _writer.writeUint8(0xd9);
      _writer.writeUint8(length);
    } else if (length <= 0xFFFF) {
      _writer.writeUint8(0xda);
      _writer.writeUint16(length);
    } else if (length <= 0xFFFFFFFF) {
      _writer.writeUint8(0xdb);
      _writer.writeUint32(length);
    } else {
      throw FormatError(
          "String is too long to be serialized with messagePack.");
    }
    _writer.writeBytes(encoded);
  }

  void _writeBinary(Uint8List buffer) {
    final length = buffer.length;
    if (length <= 0xFF) {
      _writer.writeUint8(0xc4);
      _writer.writeUint8(length);
    } else if (length <= 0xFFFF) {
      _writer.writeUint8(0xc5);
      _writer.writeUint16(length);
    } else if (length <= 0xFFFFFFFF) {
      _writer.writeUint8(0xc6);
      _writer.writeUint32(length);
    } else {
      throw FormatError("Data is too long to be serialized with messagePack.");
    }
    _writer.writeBytes(buffer);
  }

  void _writeIterable(Iterable iterable) {
    final length = iterable.length;

    if (length <= 0xF) {
      _writer.writeUint8(0x90 | length);
    } else if (length <= 0xFFFF) {
      _writer.writeUint8(0xdc);
      _writer.writeUint16(length);
    } else if (length <= 0xFFFFFFFF) {
      _writer.writeUint8(0xdd);
      _writer.writeUint32(length);
    } else {
      throw FormatError("Array is too big to be serialized with messagePack");
    }

    for (final item in iterable) {
      encode(item);
    }
  }

  void _writeMap(Map map) {
    final length = map.length;

    if (length <= 0xF) {
      _writer.writeUint8(0x80 | length);
    } else if (length <= 0xFFFF) {
      _writer.writeUint8(0xde);
      _writer.writeUint16(length);
    } else if (length <= 0xFFFFFFFF) {
      _writer.writeUint8(0xdf);
      _writer.writeUint32(length);
    } else {
      throw FormatError("Map is too big to be serialized with messagePack");
    }

    for (final item in map.entries) {
      encode(item.key);
      encode(item.value);
    }
  }

  bool _writeExt(dynamic object) {
    int? type = _extEncoder?.extTypeForObject(object);
    if (type != null) {
      if (type < 0) {
        throw FormatError("Negative ext type is reserved");
      }
      final encoded = _extEncoder?.encodeObject(object);
      if (encoded == null)
        throw FormatError('Unable to encode object. No Encoder specified.');

      final length = encoded.length;
      if (length == 1) {
        _writer.writeUint8(0xd4);
      } else if (length == 2) {
        _writer.writeUint8(0xd5);
      } else if (length == 4) {
        _writer.writeUint8(0xd6);
      } else if (length == 8) {
        _writer.writeUint8(0xd7);
      } else if (length == 16) {
        _writer.writeUint8(0xd8);
      } else if (length <= 0xFF) {
        _writer.writeUint8(0xc7);
        _writer.writeUint8(length);
      } else if (length <= 0xFFFF) {
        _writer.writeUint8(0xc8);
        _writer.writeUint16(length);
      } else if (length <= 0xFFFFFFFF) {
        _writer.writeUint8(0xc9);
        _writer.writeUint32(length);
      } else {
        throw FormatError("Size must be at most 0xFFFFFFFF");
      }
      this._writer.writeUint8(type);
      this._writer.writeBytes(encoded);
      return true;
    }
    return false;
  }
}
